src/utils/download.ts

```diff
@@ -1,25 +1,54 @@
 import * as fs from 'fs';
 import path from 'path';
-import axios, { AxiosResponse } from 'axios';
+import axios from 'axios';
+
+type DownloadProgressCallback = (downloadProgress: number) => void;
+type DownloadResult = { pathToFile: string; contentType: string };
 
 /**
  * Downloads file from remote HTTP[S] host and puts its contents to the
  * specified location.
  * @param url URL of the file to download.
  * @param destinationPath Path to the destination file.
+ * @param destinationFileName The file name.
  */
-export async function download(url: string, filePath: string): Promise<string> {
-	const fileName = url.split('/').pop();
-	if (!fileName) {
-		throw new Error('Invalid remote path. No file name found.');
-	}
-	try {
-		const response: AxiosResponse = await axios.get(url, { responseType: 'arraybuffer' });
-		const pathToFile = path.join(filePath, fileName);
 
-		fs.writeFileSync(pathToFile, response.data);
-		return pathToFile;
+export async function download(
+	url: string,
+	destinationPath: string,
+	destinationFileName: string,
+	downloadProgressCallback?: DownloadProgressCallback
+): Promise<DownloadResult> {
+	const pathToFile = path.join(destinationPath, destinationFileName);
+	const writer = fs.createWriteStream(pathToFile);
+	try {
+		const { data, headers } = await axios({
+			method: 'get',
+			url: url,
+			responseType: 'stream'
+		});
+		const totalLength = headers['content-length'];
+		const contentType = headers['content-type'];
+		let downloadedLength = 0;
+		data.on('data', (chunk: string | unknown[]) => {
+			downloadedLength += chunk.length;
+			downloadProgressCallback && downloadProgressCallback((downloadedLength / totalLength) * 100);
+		});
+		return new Promise((resolve) => {
+			data.pipe(writer);
```

Comment:
Were there problems using `pipeline` vs. `pipe`? I advocated strongly for the former due to the inherent difficulties in handling stream errors correctly. See here:
https://dev.to/morz/pipeline-api-the-best-way-to-handle-stream-errors-that-nobody-tells-you-about-122o